File: ute-bookstore-backend/src/main/java/com/hcmute/bookstore/Entity/ChatMessage.java
package com.hcmute.bookstore.Entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "chat_messages")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ChatMessage {

    @Id
    @Column(name = "id", length = 50)
    private String id;

    // Tên người dùng hoặc ID người dùng để định danh cuộc hội thoại
    private String userName;

    // Vai trò: "USER" hoặc "ADMIN"
    private String senderRole;

    private String receiverName;

    @Column(columnDefinition = "TEXT")
    private String content;

    // --- QUẢN LÝ MEDIA (ẢNH/VIDEO MÃ HÓA) ---
    // Lưu URL hoặc Path của file sau khi đã mã hóa
    @Column(name = "media_url", columnDefinition = "TEXT")
    private String mediaUrl;

    // Loại tin nhắn: TEXT, IMAGE, VIDEO, PRODUCT, ORDER
    @Enumerated(EnumType.STRING)
    private MessageType messageType;

    @Enumerated(EnumType.STRING)
    private MessageStatus status;

    // --- TÍNH NĂNG NÂNG CAO ---

    @Enumerated(EnumType.STRING)
    @Column(name = "reaction_type", length = 20)
    private ReactionType reaction;

    // Trả lời tin nhắn (Self-reference)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reply_to_id")
    private ChatMessage replyTo;

    // Đính kèm sản phẩm (Nếu có)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "book_id")
    private Books attachedBook;

    // Đính kèm đơn hàng (Nếu có)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    private Orders attachedOrder;

    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }
}

File: ute_bookstore_flutter/lib/chat/models/chat_message.dart
enum MessageType { TEXT, IMAGE, VIDEO, PRODUCT, ORDER }
enum SenderRole { USER, ADMIN }

class ChatMessage {
  final String? id;
  final String content;
  final String? mediaUrl;
  final MessageType messageType;
  final SenderRole senderRole;
  final DateTime createdAt;
  final String? userName; // Người gửi
  final String? receiverName; // Người nhận
  final String? reaction;

  // -- REPLY FIELDS --
  final String? replyToId;
  final String? replyToContent;
  final String? replyToMediaUrl;
  final MessageType? replyToMessageType;
  final String? replyToSender;

  ChatMessage({
    this.id,
    required this.content,
    this.mediaUrl,
    required this.messageType,
    required this.senderRole,
    required this.createdAt,
    this.userName,
    this.receiverName,
    this.reaction,
    this.replyToId,
    this.replyToContent,
    this.replyToMediaUrl,
    this.replyToMessageType,
    this.replyToSender,
  });

  factory ChatMessage.fromMap(Map<String, dynamic> map) {
    return ChatMessage(
      id: map['id']?.toString(),
      content: map['content'] ?? '',
      mediaUrl: map['mediaUrl'],
      senderRole: SenderRole.values.firstWhere(
        (e) => e.name.toUpperCase() == (map['senderRole']?.toString().toUpperCase() ?? 'USER'),
        orElse: () => SenderRole.USER,
      ),
      messageType: MessageType.values.firstWhere(
        (e) => e.name.toUpperCase() == (map['messageType']?.toString().toUpperCase() ?? 'TEXT'),
        orElse: () => MessageType.TEXT,
      ),
      createdAt: map['createdAt'] != null
          ? DateTime.parse(map['createdAt'])
          : DateTime.now(),
      userName: map['userName'],
      receiverName: map['receiverName'],
      reaction: map['reaction'],
      replyToId: map['replyToId'],
      replyToContent: map['replyToContent'],
      replyToMediaUrl: map['replyToMediaUrl'],
      replyToMessageType: map['replyToMessageType'] != null
          ? MessageType.values.firstWhere(
              (e) => e.name.toUpperCase() == map['replyToMessageType'].toString().toUpperCase(),
              orElse: () => MessageType.TEXT)
          : null,
      replyToSender: map['replyToSender'],
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'content': content,
      'messageType': messageType.name,
      'senderRole': senderRole.name,
      'userName': userName,
      'receiverName': receiverName,
      'mediaUrl': mediaUrl,
      'replyToId': replyToId,
    };
  }
}